docs/settings.adoc
:uri-asciidoctor: http://asciidoctor.org
:icons: font
:source-highlighter: pygments
:nofooter:
link:../README.adoc[Home]

== All settings


link:commands.adoc[commands list]

=== cVimrc

* Boolean cVimrc settings are enabled with the command 'set' + <SETTING_NAME> and disabled with the command 'set' + no<SETTING_NAME> (for example, set regexp and set noregexp)
* Boolean cVimrc settings can be inversed by adding "!" to the end
* Other settings are defined with = used as a separator and are prefixed by let (for example, let hintcharacters="abc")

|===
|Name|Type|Description|Default
| experimental
| boolean
| turns on commands flagged as experimental. View link:changelog.adoc[changelog]
| false


| handlebothcwevents
| boolean
| turns on support for C-w to both close the tab AND delete the backword
| false

| ignorenumerickeys
| boolean
| counter/repeats are ignored. e.g 3j will only execute once
| false

| debug
| boolean
| prints helpful debug messages related to configuration file
| false

| mouselesshintcharacters
| string
| the mouseless hint characters as opposed to the cvim ones. first set of letters is prioritized over second until we run out of permutations (separated by `,`)
| dsafrewq,tgcx

| newtabalwaysontheright
| boolean
| new tabs are always open on the right of the current tab
| false


| automakelinks
| boolean
| detects urls and transforms them into clickable links after document has rendered
| false


| newtaburl
| string
| url when opening a new tab with `<C-t>` - defaults to blank in incognito
| false

| searchlimit
| integer
| set the amount of results displayed in the command bar
| 25

| scrollstep
| integer
| set the amount of pixels scrolled when using the scrollUp and scrollDown commands
| 70

| timeoutlen
| integer
| The amount of time to wait for a `<Leader>` mapping in milliseconds
| 1000

| fullpagescrollpercent
| integer
| set the percent of the page to be scrolled by when using the scrollFullPageUp and scrollFullPageDown commands
| 0

| typelinkhintsdelay
| integer
| the amount of time (in milliseconds) to wait before taking input after opening a link hint with typelinkhints and numerichints enabled
| 300

| scrollduration
| integer
| the duration of smooth scrolling
| 500

| vimport
| integer
| set the port to be used with the `editWithVim` insert mode command
| 8001

| zoomfactor
| integer / double
| the step size when zooming the page in/out
| 0.1

| scalehints
| boolean
| animate link hints as they appear
| false

| hud
| boolean
| show the heads-up-display
| true

| regexp
| boolean
| use regexp in find mode
| true

| ignorecase
| boolean
| ignore search case in find mode
| true

| linkanimations
| boolean
| show fade effect when link hints open and close
| false

| numerichints
| boolean
| use numbers for link hints instead of a set of characters
| false

| dimhintcharacters
| boolean
| dim letter matches in hint characters rather than remove them from the hint
| true

| defaultnewtabpage
| boolean
| use the default chrome://newtab page instead of a blank page
| false

| cncpcompletion
| boolean
| use `<C-n>` and `<C-p>` to cycle through completion results (requires you to set the nextCompletionResult keybinding in the chrome://extensions page (bottom right)
| false

| smartcase
| boolean
| case-insensitive find mode searches except when input contains a capital letter
| true

| incsearch
| boolean
| begin auto-highlighting find mode matches when input length is greater thant two characters
| true

| typelinkhints
| boolean
| (numerichints required) type text in the link to narrow down numeric hints
| false

| autohidecursor
| boolean
| hide the mouse cursor when scrolling (useful for Linux, which doesn't auto-hide the cursor on keydown)
| false

| autofocus
| boolean
| allows websites to automatically focus an input box when they are first loaded
| true

| insertmappings
| boolean
| use insert mappings to navigate the cursor in text boxes (see bindings below)
| true

| smoothscroll
| boolean
| use smooth scrolling
| false

| autoupdategist
| boolean
| if a GitHub Gist is used to sync settings, pull updates every hour (and when Chrome restarts)
| false

| nativelinkorder
| boolean
| Open new tabs like Chrome does rather than next to the currently opened tab
| false

| showtabindices
| boolean
| Display the tab index in the tab's title
| false

| sortlinkhints
| boolean
| Sort link hint lettering by the link's distance from the top-left corner of the page
| false

| localconfig
| boolean
| Read the cVimrc config from `configpath` (when this is set, you connot save from cVim's options page
| false

| completeonopen
| boolean
| Automatically show a list of command completions when the command bar is opened
| false

| configpath
| string
| Read the cVimrc from this local file when configpath is set
| ""

| changelog
| boolean
| Auto open the changelog when cVim is updated
| true

| completionengines
| array of strings
| use only the specified search engines
| ["google", "duckduckgo", "wikipedia", "amazon"]

| blacklists
| array of strings
| disable cVim on the sites matching one of the patterns
| []

| mapleader
| string
| The default `<Leader>` key
| \

| defaultengine
| string
| set the default search engine
| "google"

| locale
| string
| set the locale of the site being completed/searched on (see example configuration below)
| ""

| homedirectory
| string
| the directory to replace `~` when using the `file` command
| ""

| qmark &lt;alphanumeric charcter&gt;
| string
| add a persistent QuickMark (e.g. ```let qmark a = ["http://google.com", "http://reddit.com"]```)
| none

| previousmatchpattern
| string (regexp)
| the pattern looked for when navigating a page's back button
| ((?!last)(prev(ious)?&#124;newer&#124;back&#124;«&#124;less&#124;&lt;&#124;‹&#124; )+)

| nextmatchpattern
| string (regexp)
| the pattern looked for when navigation a page's next button
| ((?!first)(next&#124;older&#124;more&#124;&gt;&#124;›&#124;»&#124;forward&#124; )+)

| hintcharacters
| string (alphanumeric)
| set the default characters to be used in link hint mode
| "asdfgqwertzxcvb"

| barposition
| string ["top", "bottom"]
| set the default position of the command bar
| "top"


| langmap
| string
| set a list of characters to be remapped (see vims langmap)
| ""




CHANGELOG.adoc
include::docs/changelog.adoc[]

background_scripts/options.js
var storageMethod = 'local',
  settings = {},
  Options = {};

var defaultSettings = {
  experimental: false,
  handlebothcwevents: true,
  searchlimit: 1000,
  scrollstep: 70,
  automakelinks: false,
  fullpagescrollpercent: 0,
  typelinkhintsdelay: 300,
  qmarks: { h: ['https://www.github.com'] }, // alphanumeric charactor
  sites: {},
  searchengines: {},
  searchaliases: {},
  debug: false,
  newtabalwaysontheright: false,
  hud: true,
  regexp: true,
  scalehints: false,
  linkanimations: false,
  sortlinkhints: false,
  ignorecase: true,
  numerichints: false,
  cncpcompletion: false,
  smartcase: true,
  incsearch: true,
  autohidecursor: false,
  typelinkhints: false,
  autofocus: true,
  insertmappings: true,
  defaultnewtabpage: true,
  dimhintcharacters: true,
  ignorenumerickeys: false,
  smoothscroll: true,
  autoupdategist: false,
  nativelinkorder: false,
  showtabindices: true,
  changelog: true,
  localconfig: false,
  completeonopen: false,
  debugcss: false, // Always use default COMMANDBARCSS
  scrollduration: 500,
  zoomfactor: 0.1,
  configpath: '',
  locale: '',
  mapleader: '\\',
  timeoutlen: 1000,
  newtaburl: 'https://www.google.com',
  defaultengine: 'google',
  hintcharacters: 'asdfgqwertzxcvb',
  mouselesshintcharacters: 'dsafrewq,tgcx',
  homedirectory: '',
  langmap: '',
  completionengines: ['google', 'bing', 'duckduckgo', 'wikipedia', 'amazon', 'baidu'],
  nextmatchpattern: '((?!first)(next|older|more|>|›|»|forward| )+)',
  previousmatchpattern: '((?!last)(prev(ious)?|newer|back|«|less|<|‹| )+)',
  barposition: 'top',
  vimport: 8001,
  blacklists: [],
  RC: '',
  MAPPINGS: '',
  GISTURL: '',
  FUNCTIONS: {},
  domainStylesheets: {},
  COMMANDBARCSS:
  '#cVim-command-bar,\n#cVim-command-bar-mode,\n#cVim-command-bar-input,\n#cVim-command-bar-search-results,\n.cVim-completion-item,\n.cVim-completion-item .cVim-full,\n.cVim-completion-item .cVim-left,\n.cVim-completion-item .cVim-right {\n  font-family: Helvetica, Helvetica Neue, Neue, sans-serif, monospace, Arial;\n  font-size: 10pt !important;\n  -webkit-font-smoothing: antialiased !important;\n}\n\n#cVim-command-bar {\n  position: fixed;\n  z-index: 2147483646;\n  background-color: #1b1d1e;\n  color: #bbb;\n  display: none;\n  box-sizing: content-box;\n  box-shadow: 0 3px 3px rgba(0, 0, 0, 0.4);\n  left: 0;\n  width: 100%;\n  height: 20px;\n}\n\n#cVim-command-bar-mode {\n  display: inline-block;\n  vertical-align: middle;\n  box-sizing: border-box;\n  padding-left: 2px;\n  height: 100%;\n  width: 10px;\n  padding-top: 2px;\n  color: #888;\n}\n\n#cVim-command-bar-input {\n  background-color: #1b1d1e;\n  color: #bbb;\n  height: 100%;\n  right: 0;\n  top: 0;\n  width: calc(100% - 10px);\n  position: absolute;\n}\n\n#cVim-command-bar-search-results {\n  position: fixed;\n  width: 100%;\n  max-height: 400px;\n  overflow: auto;\n  z-index: 2147483647;\n  left: 0;\n  box-shadow: 0 3px 3px rgba(0, 0, 0, 0.4);\n  background-color: #1c1c1c;\n}\n\n.cVim-completion-item,\n.cVim-completion-item .cVim-full,\n.cVim-completion-item .cVim-left,\n.cVim-completion-item .cVim-right {\n  text-overflow: ellipsis;\n  padding: 1px;\n  display: inline-block;\n  box-sizing: border-box;\n  vertical-align: middle;\n  overflow: hidden;\n  white-space: nowrap;\n}\n\n.cVim-completion-item:nth-child(even) {\n  background-color: #1f1f1f;\n}\n\n.cVim-completion-item {\n  width: 100%;\n  left: 0;\n  color: #bcbcbc;\n}\n\n.cVim-completion-item[active] {\n  width: 100%;\n  left: 0;\n  color: #1b1d1e;\n  background-color: #f1f1f1;\n}\n\n.cVim-completion-item[active] span {\n  color: #1b1d1e;\n}\n\n.cVim-completion-item .cVim-left {\n  color: #fff;\n  width: 37%;\n}\n\n.cVim-completion-item .cVim-right {\n  font-style: italic;\n  color: #888;\n  width: 57%;\n}\n\n#cVim-link-container,\n.cVim-link-hint,\n#cVim-hud,\n#cVim-status-bar {\n  font-family: Helvetica, Helvetica Neue, Neue, sans-serif, monospace, Arial;\n  font-size: 10pt !important;\n  -webkit-font-smoothing: antialiased !important;\n}\n\n#cVim-link-container {\n  /*position: absolute;*/\n  pointer-events: none;\n  width: 100%;\n  left: 0;\n  height: 100%;\n  top: 0;\n  z-index: 2147483647;\n}\n\n.cVim-link-hint {\n  position: absolute;\n  color: #302505 !important;\n  background-color: #ffd76e !important;\n  border-radius: 2px !important;\n  padding: 2px !important;\n  font-size: 8pt !important;\n  font-weight: 500 !important;\n  text-transform: uppercase !important;\n  border: 1px solid #ad810c;\n  display: inline-block !important;\n  vertical-align: middle !important;\n  text-align: center !important;\n  box-shadow: 2px 2px 1px rgba(0, 0, 0, 0.25) !important;\n}\n\n.cVim-link-hint_match {\n  color: #777;\n  text-transform: uppercase !important;\n}\n\n#cVim-hud {\n  background-color: rgba(28, 28, 28, 0.9);\n  position: fixed !important;\n  transition: right 0.2s ease-out;\n  z-index: 24724289;\n}\n\n#cVim-hud span {\n  padding: 2px;\n  padding-left: 4px;\n  padding-right: 4px;\n  color: #8f8f8f;\n  font-size: 10pt;\n}\n\n#cVim-frames-outline {\n  position: fixed;\n  width: 100%;\n  height: 100%;\n  left: 0;\n  top: 0;\n  right: 0;\n  z-index: 9999999999;\n  box-sizing: border-box;\n  border: 3px solid yellow;\n}\n'
};

chrome.storage.onChanged.addListener(function(changes) {
  if (!changes.hasOwnProperty('sessions')) {
    settings = changes.settings ? changes.settings.newValue : defaultSettings;
  }
});

Options.refreshSettings = function(callback) {
  for (var key in defaultSettings) {
    if (settings[key] === void 0) {
      settings[key] = defaultSettings[key];
    }
  }
  if (callback) {
    callback();
  }
};

Options.saveSettings = function(request) {
  settings = request.settings;
  for (var key in settings.qmarks) {
    Quickmarks[key] = settings.qmarks[key];
  }
  this.refreshSettings(function() {
    chrome.storage[storageMethod].set({ settings: settings });
    if (request.sendSettings) {
      Options.sendSettings();
    }
  });
};

Options.sendSettings = function() {
  activePorts.forEach(function(port) {
    port.postMessage({
      type: 'sendSettings',
      settings: settings
    });
  });
};

Options.getSettings = function(request, sender) {
  this.refreshSettings(function() {
    chrome.tabs.sendMessage(sender.tab.id, {
      action: 'sendSettings',
      settings: request.reset ? defaultSettings : settings
    });
  });
};

Options.setDefaults = function() {
  settings = defaultSettings;
  this.saveSettings(settings);
};

Options.getDefaults = function(request, sender, callback) {
  callback(defaultSettings);
};

// retrieve default and current settings
Options.getAllSettings = function(request, sender, callback) {
  callback({
    defaults: defaultSettings,
    current: settings
  });
};

Options.updateBlacklistsMappings = function() {
  var rc = Utils.compressArray(settings.RC.split(/\n+/)),
    i,
    index,
    line;
  if (settings.BLACKLISTS) {
    settings.blacklists = settings.BLACKLISTS.split(/\n+/);
    delete settings.BLACKLISTS;
  }
  for (i = 0; i < rc.length; ++i) {
    if (/ *let *blacklists *= */.test(rc[i])) {
      rc.splice(i, 1);
      index = i;
    }
  }
  settings.blacklists = Utils.uniqueElements(settings.blacklists);
  if (settings.blacklists.length) {
    line = 'let blacklists = ' + JSON.stringify(settings.blacklists);
    if (index) {
      rc = rc
        .slice(0, index)
        .concat(line)
        .concat(rc.slice(index));
    } else {
      rc.push(line);
    }
  }
  settings.RC = rc.join('\n');
  Options.saveSettings({ settings: settings });
};

Options.fetchGist = function() {
  httpRequest({
    url:
      settings.GISTURL +
      (settings.GISTURL.indexOf('raw') === -1 && settings.GISTURL.indexOf('github') !== -1
        ? '/raw'
        : '')
  }).then(function(res) {
    var updated;
    try {
      updated = RCParser.parse(res);
    } catch (e) {
      console.error('cVim Error: error parsing config file');
    }
    updated.GISTURL = settings.GISTURL;
    updated.COMMANDBARCSS = settings.COMMANDBARCSS;
    Options.saveSettings({
      settings: updated,
      sendSettings: true
    });
    if (updated.autoupdategist) {
      window.setTimeout(Options.fetchGist, 1000 * 60 * 60);
    }
  });
};

chrome.storage[storageMethod].get(
  'settings',
  function(data) {
    if (data.settings) {
      settings = data.settings;
      Quickmarks = settings.qmarks;
    }
    if (settings.debugcss) settings.COMMANDBARCSS = defaultSettings.COMMANDBARCSS;
    this.refreshSettings();
    this.updateBlacklistsMappings();
    if (settings.autoupdategist && settings.GISTURL) {
      this.fetchGist();
    }
  }.bind(Options)
);

chrome.runtime.onMessage.addListener(function(request, sender, callback) {
  if (Options.hasOwnProperty(request.action)) {
    Options[request.action](request, sender, callback);
  }
});
